src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs

```diff
@@ -62,52 +71,195 @@ public void ScopeDisposed(int serviceProviderHashCode, int scopedServicesResolve
             }
         }
 
         [NonEvent]
-        public void CallSiteBuilt(Type serviceType, ServiceCallSite callSite)
+        public void CallSiteBuilt(ServiceProvider provider, Type serviceType, ServiceCallSite callSite)
         {
             if (IsEnabled(EventLevel.Verbose, EventKeywords.All))
             {
                 string format = CallSiteJsonFormatter.Instance.Format(callSite);
                 int chunkCount = format.Length / MaxChunkSize + (format.Length % MaxChunkSize > 0 ? 1 : 0);
 
+                int providerHashCode = provider.GetHashCode();
                 for (int i = 0; i < chunkCount; i++)
                 {
                     CallSiteBuilt(
                         serviceType.ToString(),
-                        format.Substring(i * MaxChunkSize, Math.Min(MaxChunkSize, format.Length - i * MaxChunkSize)), i, chunkCount);
+                        format.Substring(i * MaxChunkSize, Math.Min(MaxChunkSize, format.Length - i * MaxChunkSize)), i, chunkCount,
+                        providerHashCode);
                 }
             }
         }
 
         [NonEvent]
-        public void DynamicMethodBuilt(Type serviceType, int methodSize)
+        public void DynamicMethodBuilt(ServiceProvider provider, Type serviceType, int methodSize)
         {
             if (IsEnabled(EventLevel.Verbose, EventKeywords.All))
             {
-                DynamicMethodBuilt(serviceType.ToString(), methodSize);
+                DynamicMethodBuilt(serviceType.ToString(), methodSize, provider.GetHashCode());
             }
         }
 
         [NonEvent]
-        public void ServiceRealizationFailed(Exception exception)
+        public void ServiceRealizationFailed(Exception exception, int serviceProviderHashCode)
         {
             if (IsEnabled(EventLevel.Error, EventKeywords.All))
             {
-                ServiceRealizationFailed(exception.ToString());
+                ServiceRealizationFailed(exception.ToString(), serviceProviderHashCode);
+            }
+        }
+
+        [NonEvent]
+        public void ServiceProviderBuilt(ServiceProvider provider)
+        {
+            lock (_providers)
+            {
+                _providers.Add(provider);
+            }
+
+            WriteServiceProviderBuilt(provider);
+        }
+
+        [NonEvent]
+        public void ServiceProviderDisposed(ServiceProvider provider)
+        {
+            lock (_providers)
+            {
+                _providers.Remove(provider);
+            }
+        }
+
+        [NonEvent]
+        private void WriteServiceProviderBuilt(ServiceProvider provider)
+        {
+            if (IsEnabled(EventLevel.Informational, Keywords.ServiceProviderInitialized))
+            {
+                int singletonServices = 0;
+                int scopedServices = 0;
+                int transientServices = 0;
+
+                StringBuilder descriptorBuilder = new StringBuilder("{ \"descriptors\":[ ");
```

Comment:
Generally speaking, the pattern is to do less string formatting and to do more writing of events with payloads that can then be processed out-of-proc.  This looks like it might be a place where building up the state in-proc is preferred, or that you have to string-ify things anyway, but just wanted to call that out.  Definitely good that this is under its own keyword.

Reply:
Yeah, here we want to dump a bunch of data - the list of registered DI services. We are doing something similar above already:

https://github.com/dotnet/runtime/blob/a1aeb132ca60becb1e8ffe02150d59782c1214c2/src/libraries/Microsoft.Extensions.DependencyInjection/src/DependencyInjectionEventSource.cs#L80-L94